sdk/couchbase-core/src/kvclient.rs

```diff
@@ -0,0 +1,338 @@
+            Some(GetErrorMapRequest { version: 2 })
+        } else {
+            None
+        };
+
+        let bootstrap_auth = if let Some(ref auth) = config.authenticator {
+            let creds = auth.get_credentials(ServiceType::Memd, config.address.to_string())?;
+
+            Some(SASLAuthAutoOptions {
+                username: creds.username.clone(),
+                password: creds.password.clone(),
+                enabled_mechs: vec![AuthMechanism::ScramSha512, AuthMechanism::ScramSha256],
+            })
+        } else {
+            None
+        };
+
+        let bootstrap_select_bucket =
+            config
+                .selected_bucket
+                .as_ref()
+                .map(|bucket_name| SelectBucketRequest {
+                    bucket_name: bucket_name.clone(),
+                });
+
+        let should_bootstrap = boostrap_hello.is_some()
+            || bootstrap_auth.is_some()
+            || bootstrap_get_error_map.is_some();
+
+        if should_bootstrap && config.disable_bootstrap {
+            // TODO: error model needs thought.
+            return Err(CoreError {
+                msg: "oopsies".to_string(),
+            });
+        }
+
+        let (connection_close_tx, mut connection_close_rx) =
+            oneshot::channel::<crate::memdx::client::Result<()>>();
+        let memdx_client_opts = DispatcherOptions {
+            on_connection_close_handler: Some(connection_close_tx),
+            orphan_handler: opts.orphan_handler,
+        };
+
+        let closed = Arc::new(AtomicBool::new(false));
+        let closed_clone = closed.clone();
+
+        tokio::spawn(async move {
+            // There's not much to do when the connection closes so just mark us as closed.
+            if connection_close_rx.await.is_ok() {
+                closed_clone.store(true, Ordering::Relaxed)
+            };
+        });
+
+        let conn = Connection::connect(
+            config.address,
+            ConnectOptions {
+                tls_config: None,
+                deadline: Instant::now().add(Duration::new(7, 0)),
+            },
+        )
+        .await?;
+
+        let remote_addr = match conn.peer_addr() {
+            Some(addr) => *addr,
+            None => config.address,
+        };
+
+        let local_addr = *conn.local_addr();
+
+        // let mut cli = Client::new(conn, memdx_client_opts);
+
+        let mut cli = D::new(conn, memdx_client_opts);
+
+        let mut kv_cli = StdKvClient {
+            remote_addr,
+            local_addr,
+            pending_operations: 0,
+            cli,
+            current_config: Mutex::new(config),
+            supported_features: vec![],
+            selected_bucket: Arc::new(Mutex::new(None)),
+            closed,
+        };
+
+        if should_bootstrap {
+            if let Some(b) = &bootstrap_select_bucket {
+                let mut guard = kv_cli.selected_bucket.lock().unwrap();
```

Comment:
NIT: More of a general thing, and we don't have to change it now, but should we prefer .expect() (w/ appropriate panic message) over .unwrap()?

Reply:
We shouldn't really be using either anywhere that we can proceed without a panic. In this case acquiring the lock only returns an error if it's been poisoned, in which case there was a panic elsewhere and we actually can't do much other than panic.